### ISSUE TITLE
No source information available

### ISSUE THREAD

**Issue by ayushkarnawat** [google/pprof#647]
### What version of pprof are you using?

Latest head: 02619b876842e0d0afb5e5580d3a374dad740edb. Note: I am building the git revision within a `nix` derivation, but even trying with `go get -u github.com/google/pprof` results in the same "error". 

### What operating system and processor architecture are you using?

System: `"x86_64-darwin"`
OS: `Darwin 20.5.0, macOS 10.16`

### What did you do?

Following [this issue](https://github.com/gperftools/gperftools/issues/956) for the test case; reproduced (with minor changes) below for completeness. Found similar behavior for other files, but (legally) unable to share those examples. In a folder, create

```c
// lib.cpp
long long int factorial(long long int m_nValue) {
  long long int result=m_nValue;
  long long int result_next;
  long long int pc = m_nValue;
  do {
    result_next = result*(pc-1);
    result = result_next;
    pc--;
  } while(pc>2);
  m_nValue = result;
  return m_nValue;
}

int foo() {
  int m = 0;
  for (int i = 1; i < 100000; ++i)
    m += factorial(i);
  return m;
}
```

```c
// main.cpp
#include <iostream>

int foo();

int main(void) {
  std::cout << "Hello\n";
  int out = foo();
  std::cout << out << std::endl;
  return EXIT_SUCCESS;
}
```

and then compile, link, and profile with the following flags:

```
clang++ -g -fno-omit-frame-pointer -dynamic-lib -o liblib.dylib lib.cpp
clang++ -g -fno-omit-frame-pointer -c main.cpp -o main.o
clang++ -o main main.o -L. -llib -lprofiler
CPUPROFILE=main.prof ./main
PPROF_BINARY_PATH=./main
pprof -list ".*" main.prof
```

to generate the following profile: [main.prof.zip](https://github.com/google/pprof/files/7200601/main2.prof.zip)

### What did you expect to see?

Something similar to below; essentially a line by line breakdown of the profiling report, rather than at the function level. 

```c
$ pprof -list ".*" main.prof
Local symbolization failed for libdyld.dylib: unrecognized binary: /usr/lib/system/libdyld.dylib
Some binary filenames not available. Symbolization may be incomplete.
Try setting PPROF_BINARY_PATH to the search path for local binaries.
Total: 690ms
ROUTINE ========================  in ??
         0      690ms (flat, cum)   100% of Total
 Error: Could not find file ?? on path /Users/me/test
ROUTINE ======================== factorial in /Users/me/test/lib.cpp
     690ms      690ms (flat, cum)   100% of Total
         .          .      3:    long long int result=m_nValue;
         .          .      4:    long long int result_next;
         .          .      5:    long long int pc = m_nValue;
         .          .      6:    do
         .          .      7:    {
     310ms      310ms      8:        result_next = result*(pc-1);
     250ms      250ms      9:        result = result_next;
      60ms       60ms     10:        pc--;
      70ms       70ms     11:    } while(pc>2);
         .          .     12:    m_nValue = result;
         .          .     13:    return m_nValue;
         .          .     14:}
         .          .     15:
         .          .     16:int foo() {
```

### What did you see instead?

```sh
$ pprof -list '.*' main.prof
Local symbolization failed for libdyld.dylib: stat /usr/lib/system/libdyld.dylib: no such file or directory
Some binary filenames not available. Symbolization may be incomplete.
Try setting PPROF_BINARY_PATH to the search path for local binaries.
Total: 2.52s
No source information for 
No source information for foo
```

**Comment by nolanmar511** [comment]
https://github.com/google/pprof/tree/master/doc#annotated-code contains more information about annotated code output formats.

One important thing about these output formats: the source or binaries must be locally available and they must be locally available in a location which the pprof CLI can identify. There could, of course, be other issues, but this is a common enough one that it needs to be ruled out first.

It's not clear from your issue, but did you try setting `$PPROF_BINARY_PATH` as suggested in the output you saw?  Alternatively, did you try specifying the binary in the pprof command (`pprof <format> [options] [binary] <source> ...` -- the help texted outputed by `pprof -help` documents that the path to the binary can be specified this way)?

**Comment by ayushkarnawat** [comment]
Thanks @nolanmar511, I totally forgot to attach that I did set the `$PPROF_BINARY_PATH` to the local binary (following the error msg above), but got the same result. I believe the reason this error appears despite having set `$PPROF_BINARY_PATH=./main` is because on MacOS BigSur systems, Apple made changes such that dynamically linked libraries can no longer be searched as files within the system, but are directly loaded into memory when running the executable.

When trying to specify the binary in the `pprof` command directly (rather than `$PPROF_BINARY_PATH`), I get the following error message:

```sh
$ pprof -text -lines -trim=false ./main main.prof
./main: parsing profile: unrecognized profile format
Fetched 1 source profiles out of 2
Local symbolization failed for libsystem_c.dylib: stat /usr/lib/system/libsystem_c.dylib: no such file or directory
Local symbolization failed for libc++.1.dylib: stat /usr/lib/libc++.1.dylib: no such file or directory
Local symbolization failed for libdyld.dylib: stat /usr/lib/system/libdyld.dylib: no such file or directory
Some binary filenames not available. Symbolization may be incomplete.
Try setting PPROF_BINARY_PATH to the search path for local binaries.
File: libc++.1.dylib
Type: cpu
Showing nodes accounting for 2.81s, 100% of 2.81s total
      flat  flat%   sum%        cum   cum%
     2.80s 99.64% 99.64%      2.80s 99.64%  foo
     0.01s  0.36%   100%      0.01s  0.36%  [libsystem_c.dylib]
         0     0%   100%      0.01s  0.36%  <unknown>
         0     0%   100%      0.01s  0.36%  [libc++.1.dylib]
         0     0%   100%      0.01s  0.36%  [libdyld.dylib]
```

So, it is detecting the function, but unable to print detailed line by line output, even when specifying the `-lines` option.

**Comment by nolanmar511** [comment]
That would suggest the profile does not have line number (or address number) information. The pprof CLI cannot display information that is not available in the profile. 

Was the binary compiled with sufficient debug information?

**Comment by ayushkarnawat** [comment]
It seems so, the above error was with `-g` flag, will try with `-g O0` perhaps?

**Comment by ayushkarnawat** [comment]
I should note I am generating the profile using gperftools, if that would make any difference

**Comment by nolanmar511** [comment]
It looks as though ./main is not being recognized as a binary. `$PPROF_BINARY_PATH` would be used to specify a directory containing binaries; the binary in a `pprof <format> [options] [binary] <source> ...` command need to be a singular file; from your comment which it is.

**Comment by nolanmar511** [comment]
At this point in time, I'm going to close this issue since it now looks like the issue encountered seems somewhat different that the initially suspected issue, and because, with current information, I don't think there's enough information to conclude there's a bug in `pprof`.

To summarize my understanding: the pprof CLI is able to symbolize the profile (because "foo" appears in the output, rather than something in square brackets), but there are no line numbers (which suggests that information is not available in the binary). I'll call out pprof's symbolization options (`-symbolize=none` to disable symbolization; `-symbolize= force` to force resymbolization), in case those are useful.

**Comment by ayushkarnawat** [comment]
TL;DR: Root problem was with the clang version packaged within a `nix` environment. It does not generate a valid UUID for the binary, which is necessary to associate the DWARF debug symbols to their respective line in the `src` code. 

To follow up, I found the source of the error after a lot of reading and experimentation. As suspected above, the error was not actually an issue within `pprof`, but rather when building the binary within a **nix environment** using clang/llvm. As you will notice, the key issue was using the `nix`-provided clang binary rather than one installed via `brew`. For completeness, I explain what is happening, in case anyone encounters this in the future (however unlikely).

Digging deeper into why I'm unable to get proper symbolization for the binary, I [found](https://stackoverflow.com/a/12827463/7148714) that the debug information is not actually contained in the binary on Apple systems. Rather, the main executable (symbolically) links to the debug information present in each individual compiled object file (if debug symbols are enabled). This means that even though the debug info is present, it is not contained in the final binary itself. We can confirm this by checking the section headers of each file using `objdump -h <path/to/binary>`:

```sh
$ objdump -h main
main:	file format Mach-O 64-bit x86-64

Sections:
Idx Name             Size     VMA              Type
  0 __text           00000acf 0000000100001250 TEXT
  1 __stubs          0000007e 0000000100001d20 TEXT
  2 __stub_helper    000000ce 0000000100001da0 TEXT
  3 __gcc_except_tab 00000080 0000000100001e70 DATA
  4 __cstring        00000007 0000000100001ef0 DATA
  5 __unwind_info    00000108 0000000100001ef8 DATA
  6 __nl_symbol_ptr  00000008 0000000100002000 DATA
  7 __got            00000030 0000000100002008 DATA
  8 __la_symbol_ptr  000000a8 0000000100002038 DATA
```

```sh
$ objdump -h main.o
main.o:	file format Mach-O 64-bit x86-64

Sections:
Idx Name             Size     VMA              Type
  0 __text           00000a7e 0000000000000000 TEXT
  1 __gcc_except_tab 00000080 0000000000000a80 DATA
  2 __cstring        00000007 0000000000000b00 DATA
  3 __debug_loc      0000224a 0000000000000b07 DATA
  4 __debug_abbrev   00000769 0000000000002d51 DATA
  5 __debug_info     00008016 00000000000034ba DATA
  6 __debug_ranges   00000090 000000000000b4d0 DATA
  7 __debug_str      00009d13 000000000000b560 DATA
  8 __apple_names    00000e2c 0000000000015273 DATA
  9 __apple_objc     00000024 000000000001609f DATA
 10 __apple_namespac 000000ac 00000000000160c3 DATA
 11 __apple_types    000012d0 000000000001616f DATA
 12 __compact_unwind 00000a00 0000000000017440 DATA
 13 __eh_frame       00000ce8 0000000000017e40 DATA
 14 __debug_line     00000c6a 0000000000018b28 DATA

$ objdump -h lib.o
lib.o:	file format Mach-O 64-bit x86-64

Sections:
Idx Name             Size     VMA              Type
  0 __text           0000005f 0000000000000000 TEXT
  1 __debug_str      0000011c 000000000000005f DATA
  2 __debug_loc      0000011e 000000000000017b DATA
  3 __debug_abbrev   0000007d 0000000000000299 DATA
  4 __debug_info     000000f9 0000000000000316 DATA
  5 __apple_names    00000090 000000000000040f DATA
  6 __apple_objc     00000024 000000000000049f DATA
  7 __apple_namespac 00000024 00000000000004c3 DATA
  8 __apple_types    00000066 00000000000004e7 DATA
  9 __compact_unwind 00000040 0000000000000550 DATA
 10 __eh_frame       00000068 0000000000000590 DATA
 11 __debug_line     00000062 00000000000005f8 DATA
```

Notice how the compiled object files each contain the `__debug_info` section, but the final binary does not. By further checking the symbols table,

```sh
$ nm -ap main | grep OSO
0000000061682f92 - 03 0001   OSO </path/to/root>/main2.o
0000000061682f91 - 03 0001   OSO </path/to/root>/lib.o
```

we can confirm that the final binary `main` contains `OSO` stabs which point to where the object file is located, thus allowing us to read the corresponding symbols in the associated cpp source file(s) via the `dwarfdump` command. However, due to [Apple's "Lazy" DWARF scheme](http://wiki.dwarfstd.org/index.php?title=Apple%27s_%22Lazy%22_DWARF_Scheme), to properly symbolize the instruction addresses contained in each object file to their respective line, we first have to map them to their final address locations by running `dsymutil` on the executable. This will generate a dSYM bundle, and "stamp" the same UUID from the generated executable, thus giving us a reliable way to assess if a given binary and the DWARF are associated.

So, here is where the problem lies. When building a binary using clang/llvm within a nix-based shell environment, no `LC_UUID` load command is printed when viewing the output of `otool -l main`. As such, the DWARF contained in the dSYM
bundle cannot be associated with the instruction addresses in the final binary. In fact, viewing the UUID using `dwarfdump -u main.dSYM/` prints `UUID: 00000000-0000-0000-0000-000000000000 (x86_64) main.dSYM/Contents/Resources/DWARF/main`, which is invalid. Therefore, because the UUIDs don't match, `pprof` is unable to properly symbolize instruction addresses to their respective line in the source code.

Comparing this to the executable built using the `brew` installed version of clang/llvm, a proper UUID is generated:

```sh
$ otool -l main
Load command 8
     cmd LC_UUID
 cmdsize 24
    uuid 198BC9BF-DEBF-3D1E-A6D2-FDFF46E4BFEB
```

This suggests that there is something wrong occurring with the clang version wrapped in `nixpkgs`.
